connect-inject/health_check_handler.go

```diff
@@ -0,0 +1,317 @@
+			Notes:                          reason,
+			TLSSkipVerify:                  true,
+			SuccessBeforePassing:           1,
+			FailuresBeforeCritical:         1,
+			DeregisterCriticalServiceAfter: "",
+		},
+		// For Consul-ENT only, this will be omitted for OSS, aka ""
+		Namespace: pod.Annotations[annotationConsulDestinationNamespace],
+	})
+	if err != nil {
+		t.Log.Error("unable to register health check with Consul from k8s: %v", err)
+		return err
+	}
+	return nil
+}
+
+// setConsulHealthCheckStatus will update the TTL status of the check
+func (t *HealthCheckHandler) setConsulHealthCheckStatus(healthCheckID, reason string, fail bool) error {
+	if fail == true {
+		// TODO: Use UpdateTTL() instead for both of these
+		return t.Client.Agent().FailTTL(healthCheckID, reason)
+	} else {
+		return t.Client.Agent().PassTTL(healthCheckID, reason)
+	}
+}
+
+// Init handles any handler initialization and is a no-op
+func (t *HealthCheckHandler) Init() error {
+	t.Log.Debug("handlerInit starting")
+	return t.Reconcile()
+}
+
+// Reconcile iterates through all Pods with the appropriate label and compares the
+// current health check status against that which is stored in Consul and updates
+// the consul health check accordingly. if the health check doesnt yet exist it will
+func (t *HealthCheckHandler) Reconcile() error {
+	podList, err := t.Clientset.CoreV1().Pods(corev1.NamespaceAll).List(ctx.Background(),
+		metav1.ListOptions{LabelSelector: labelInject})
+	if err != nil {
+		t.Log.Error("unable to get pods failing handler, err:%v", err)
+		return err
+	}
+	for _, pod := range podList.Items {
+		t.Log.Debug("processing Pod %v", pod.Name)
+		if pod.Status.Phase != corev1.PodRunning {
+			t.Log.Info("pod %v is not running", pod.Name, pod.Status.Phase)
+			continue
+		}
+		// 1. Fetch the health checks for this agent
+		err = t.updateConsulClient(&pod)
+		if err != nil {
+			t.Log.Error("unable to set client connection for %v", pod.Name)
+			continue
+		}
+		serviceID := t.getConsulServiceID(&pod)
+		healthCheckID := t.getConsulHealthCheckID(&pod)
+
+		// 2. Fetch the health checks on this agent related to this healthCheckID
+		filter := "Name == `" + healthCheckID + "`"
+		checks, err := t.Client.Agent().ChecksWithFilter(filter)
+		if err != nil {
+			t.Log.Error("unable to get agent health checks for %v", healthCheckID, filter)
+			continue
+		}
+		if checks == nil || len(checks) == 0 {
+			// TODO we need to handle this
+			t.Log.Error("======== we really shouldn't be here!, %v", checks)
+			continue
+		}
+		status, reason, err := t.getReadyStatusAndReason(&pod)
+		if err != nil {
+			t.Log.Error("unable to get pod status: %v", err)
+			continue
+		}
+		initialStatus := healthCheckPassing
+		if _, ok := checks[healthCheckID]; !ok {
+			// 3. If the check doesnt exist yet, create it!
+			if status == corev1.ConditionFalse {
+				initialStatus = healthCheckCritical
+			} else {
+				// no need to store a reason if passing
+				reason = ""
+			}
```

Comment:
We might want to flip this to `status != corev1.ConditionTrue` 
A condition can either be `true`, `false` or `unknown`. I think we should check if it is not true, so that we respond to condition unknown as fail instead of successful